logicalplan/propagate_selectors.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"sort"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// PropagateMatchersOptimizer implements matcher propagation between
+// two vector selectors in a binary expression.
+type PropagateMatchersOptimizer struct{}
+
+func (m PropagateMatchersOptimizer) Optimize(expr parser.Expr) parser.Expr {
+	traverse(&expr, func(expr *parser.Expr) {
+		binOp, ok := (*expr).(*parser.BinaryExpr)
+		if !ok {
+			return
+		}
+
+		// The optimizer cannot be applied to comparison operations.
+		if binOp.Op.IsComparisonOperator() {
+			return
+		}
+
+		// TODO(fpetkovski): Investigate support for vector matching on a subset of labels.
+		if binOp.VectorMatching != nil && len(binOp.VectorMatching.MatchingLabels) > 0 {
+			return
+		}
+
+		// TODO(fpetkovski): Investigate support for one-to-many and many-to-one.
+		if binOp.VectorMatching != nil && binOp.VectorMatching.Card != parser.CardOneToOne {
+			return
+		}
+
+		propagateMatchers(binOp)
+	})
+
+	return expr
+}
+
+func propagateMatchers(binOp *parser.BinaryExpr) {
+	lhSelector, ok := binOp.LHS.(*parser.VectorSelector)
+	if !ok {
+		return
+	}
+	rhSelector, ok := binOp.RHS.(*parser.VectorSelector)
+	if !ok {
+		return
+	}
+	// This case is handled by MergeSelectsOptimizer.
+	if lhSelector.Name == rhSelector.Name {
```

Comment:
What about queries like `node_filesystem_files{host="$host", mountpoint="/"} - node_filesystem_files_free / node_filesystem_files_free`? Basically LHS or RHS could be another vector

Reply:
I tried to cover this case but it's not as simple to implement. Maybe we can iterate on it in a separate PR?